kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -75,8 +80,10 @@ public class AsyncProcessor<KIn, VIn, KOut, VOut>
   private final Map<String, AbstractAsyncStoreBuilder<?, ?, ?>> connectedStoreBuilders;
 
   // Tracks all pending events, ie from moment of creation to end of finalization/transition
-  // to "DONE" state. Used to make sure all events are flushed during a commit
-  private final Set<AsyncEvent> pendingEvents = new HashSet<>();
+  // to "DONE" state. Used to make sure all events are flushed during a commit.
+  // We use a concurrent map here so metrics readers can query its size. Otherwise, only
+  // the stream thread should access this.
+  private final Map<AsyncEvent, Object> pendingEvents = new ConcurrentHashMap<>();
```

Comment:
yeah that should be the only effect practically - you might get a cached value for `size` so it might flip flop depending on where the metric is being read from (its an int so I don't think you can get a partial read). That said, `HashSet` is technically free to throw if multiple threads access it. It just feels safer to use a concurrent map if it's going to be accessed concurrently.

Reply:
> That said, HashSet is technically free to throw if multiple threads access it.

Ah, didn't know that. Anyways that all makes sense 👍